File: algorithms/0695-max-area-of-island/Solution.java
// Max Area of Island - DFS flood-fill; return area, track global max.
// Time O(m*n), Space O(m*n) worst-case recursion depth.

class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        if (grid == null || grid.length == 0) return 0;
        int maxArea = 0;
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (grid[r][c] == 1) {
                    maxArea = Math.max(maxArea, dfs(grid, r, c));
                }
            }
        }
        return maxArea;
    }

    private int dfs(int[][] grid, int r, int c) {
        if (r < 0 || c < 0 || r >= grid.length || c >= grid[0].length) return 0;
        if (grid[r][c] != 1) return 0;
        grid[r][c] = 0;
        return 1
            + dfs(grid, r + 1, c)
            + dfs(grid, r - 1, c)
            + dfs(grid, r, c + 1)
            + dfs(grid, r, c - 1);
    }
}

File: algorithms/0695-max-area-of-island/Solution.kt
// Max Area of Island - DFS flood-fill; return area, track global max.
// Time O(m*n), Space O(m*n) worst-case recursion depth.

class Solution {
    fun maxAreaOfIsland(grid: Array<IntArray>): Int {
        if (grid.isEmpty()) return 0
        var maxArea = 0
        for (r in grid.indices) {
            for (c in grid[0].indices) {
                if (grid[r][c] == 1) {
                    maxArea = maxOf(maxArea, dfs(grid, r, c))
                }
            }
        }
        return maxArea
    }

    private fun dfs(grid: Array<IntArray>, r: Int, c: Int): Int {
        if (r !in grid.indices || c !in grid[0].indices) return 0
        if (grid[r][c] != 1) return 0
        grid[r][c] = 0
        return 1 +
            dfs(grid, r + 1, c) +
            dfs(grid, r - 1, c) +
            dfs(grid, r, c + 1) +
            dfs(grid, r, c - 1)
    }
}
